File: Data Structures & Algorithms/count-connected-components/submission-1.swift
class Solution {
    func countComponents(_ n: Int, _ edges: [[Int]]) -> Int {
        var nodes: [Int:Node] = [:]
        
        // Build Graphs
        for edge in edges {
            let a = edge[0]
            let b = edge[1]
            var aNode: Node
            var bNode: Node
            if nodes[a] != nil { aNode = nodes[a]! } else { aNode = Node(a) }
            if nodes[b] != nil { bNode = nodes[b]! } else { bNode = Node(b) }
            nodes[a] = aNode
            nodes[b] = bNode

            aNode.neighbors[b] = bNode
            bNode.neighbors[a] = aNode
        }

        let rem = n-nodes.count

        var graphCount = 0
        while nodes.count > 0 {
            graphCount+=1
            let firstNode = nodes.first!.value
            var visited: Set<Int> = []
            dfs(firstNode, nil, &visited)
            for visit in visited {
                nodes.removeValue(forKey: visit)
            }
        }
        
        return graphCount+rem


        func dfs(_ node: Node, _ prev: Node?, _ visited: inout Set<Int>) {
            if visited.contains(node.val) { return }
            visited.insert(node.val)

            for (_, neighbor) in node.neighbors {
                dfs(neighbor, node, &visited)
            }
        }
    }
}

class Node {
    let val: Int
    var neighbors: [Int: Node] = [:]

    init (_ val: Int) {
        self.val = val
    }
}

File: Data Structures & Algorithms/count-paths/submission-1.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        visited = {}
        
        def dfs(i: int, j: int) -> int:
            if i < 0 or i > m : return 0
            if j < 0 or j > n : return 0
            if i == m-1 and j == n-1 : return 1
            if (i, j) in visited: return visited[(i,j)]

            sum = dfs(i+1, j) + dfs(i, j+1)
            visited[(i,j)] = sum
            return sum
        
        return dfs(0,0)